samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemDescriptor.java

```diff
@@ -0,0 +1,347 @@
+  private static final String CONSUMER_FETCH_THRESHOLD_CONFIG_KEY = KafkaConfig.CONSUMER_FETCH_THRESHOLD();
+  private static final String CONSUMER_FETCH_THRESHOLD_BYTES_CONFIG_KEY = KafkaConfig.CONSUMER_FETCH_THRESHOLD_BYTES();
+  private static final String CONSUMER_FETCH_MESSAGE_MAX_BYTES_KEY = "systems.%s.consumer.fetch.message.max.bytes";
+  private static final String CONSUMER_CONFIGS_CONFIG_KEY = "systems.%s.consumer.%s";
+  private static final String PRODUCER_BOOTSTRAP_SERVERS_CONFIG_KEY = "systems.%s.producer.bootstrap.servers";
+  private static final String PRODUCER_CONFIGS_CONFIG_KEY = "systems.%s.producer.%s";
+
+  private List<String> consumerZkConnect = Collections.emptyList();
+  private Optional<String> consumerAutoOffsetResetOptional = Optional.empty();
+  private Optional<Integer> consumerFetchThresholdOptional = Optional.empty();
+  private Optional<Long> consumerFetchThresholdBytesOptional = Optional.empty();
+  private Optional<Long> consumerFetchMessageMaxBytesOptional = Optional.empty();
+  private Map<String, String> consumerConfigs = Collections.emptyMap();
+  private List<String> producerBootstrapServers = Collections.emptyList();
+  private Map<String, String> producerConfigs = Collections.emptyMap();
+
+  /**
+   * Constructs a {@link KafkaSystemDescriptor} instance with no system level serde.
+   * Serdes must be provided explicitly at stream level when getting input or output descriptors.
+   *
+   * @param systemName name of this system
+   */
+  public KafkaSystemDescriptor(String systemName) {
+    super(systemName, FACTORY_CLASS_NAME, null);
+  }
+
+  /**
+   * Constructs a {@link KafkaSystemDescriptor} instance with {@code serde} as the system level serde.
+   *
+   * @param systemName name of this kafka system
+   * @param systemSerde default serde for the system, or null.
+   *                    If null, input/output descriptor serde must be provided at a stream level.
+   *                    A {@code KVSerde<NoOpSerde, NoOpSerde>} or {@code NoOpSerde} may be provided if the
+   *                    System's consumer deserializes the incoming messages itself, and no further deserialization
+   *                    is required from the framework.
+   */
+  public KafkaSystemDescriptor(String systemName, Serde<SystemMessageType> systemSerde) {
+    super(systemName, FACTORY_CLASS_NAME, systemSerde);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public KafkaInputDescriptor<SystemMessageType> getInputDescriptor(String streamId) {
+    if (!getSystemSerde().isPresent()) {
+      throw new SamzaException(
+          String.format("System: %s does not have a system level serde. " +
+              "Serde for input stream: %s must be specified explicitly, e.g., using " +
+                  "KafkaSystemDescriptor#getInputDescriptor(String, Serde)", getSystemName(), streamId));
+    }
+    return new KafkaInputDescriptor<>(streamId, this, getSystemSerde().get(), null);
```

Comment:
Is it possible to have full implementations in the base class, and then just have a single abstract method which returns the concrete descriptor? Otherwise, if more "getInputDescriptor" methods are added, then this needs to implement all of them, but it looks very similar to other classes.

Reply:
Good suggestion, will try it out.

Reply:
I don't think this would work, because there's not enough information at compile time to infer the  concrete type of the InputDescriptor correctly. We can experiment more offline in case I missed something.